### PR TITLE
feat(spanner): add support for external hosts

### PR DESCRIPTION
**Pull request by sagnghos**
Thank you for opening a Pull Request! Before submitting your PR, there are a few things you can do to make sure it goes smoothly:
- [x] Make sure to open an issue as a [bug/issue](https://github.com/googleapis/java-spanner/issues/new/choose) before writing your code!  That way we can discuss the change, evaluate designs, and agree on the general idea
- [x] Ensure the tests and linter pass
- [x] Code coverage does not decrease (if any source code was changed)
- [x] Appropriate docs were updated (if necessary)

Fixes #<issue_number_goes_here> ☕️

If you write sample code, please follow the [samples format](
https://github.com/GoogleCloudPlatform/java-docs-samples/blob/main/SAMPLE_FORMAT.md).

### CODE REVIEW

File: google-cloud-spanner/src/main/java/com/google/cloud/spanner/SessionClient.java [hunk 1]
```diff
@@ -49,10 +51,16 @@ private SessionId(DatabaseId db, String name) {
     static SessionId of(String name) {
       Preconditions.checkNotNull(name);
       Map<String, String> parts = NAME_TEMPLATE.match(name);
+      if (parts == null) {
+        parts = EXTERNAL_HOST_NAME_TEMPLATE.match(name);
```
Comment (krishnamoorthy-r) [inline google-cloud-spanner/src/main/java/com/google/cloud/spanner/SessionClient.java:55]: Can you add unit tests to cover this pattern and ensure the functionality works as intended?

File: google-cloud-spanner/src/main/java/com/google/cloud/spanner/DatabaseId.java [hunk 2]
```diff
@@ -94,6 +94,11 @@ public static DatabaseId of(String project, String instance, String database) {
     return new DatabaseId(new InstanceId(project, instance), database);
   }
 
+  /** Creates a {@code DatabaseId} given instance and database IDs. */
+  public static DatabaseId of(String instance, String database) {
+    return new DatabaseId(new InstanceId("default", instance), database);
```
Comment (krishnamoorthy-r) [inline google-cloud-spanner/src/main/java/com/google/cloud/spanner/DatabaseId.java:99]: I think we should document to users that this method assumes instance Id to be "default".

Comment (ko3a4ok) [reply google-cloud-spanner/src/main/java/com/google/cloud/spanner/DatabaseId.java:99]: maybe not instance Id, but rather project Id?

File: google-cloud-spanner/src/main/java/com/google/cloud/spanner/SessionClient.java (hunk 1 above)
Comment (sagnghos) [reply google-cloud-spanner/src/main/java/com/google/cloud/spanner/SessionClient.java:55]: Added unit test to cover existing pattern (for regression) as well as new pattern

File: google-cloud-spanner/src/main/java/com/google/cloud/spanner/DatabaseId.java [hunk 3]
```diff
@@ -94,6 +94,11 @@ public static DatabaseId of(String project, String instance, String database) {
     return new DatabaseId(new InstanceId(project, instance), database);
   }
 
+  /** Creates a {@code DatabaseId} with "default" project, given instance and database IDs. */
```
Comment (krishnamoorthy-r) [inline google-cloud-spanner/src/main/java/com/google/cloud/spanner/DatabaseId.java:97]: nit nit pick: "default" as project id.

File: google-cloud-spanner/src/test/java/com/google/cloud/spanner/SessionClientTests.java [hunk 4]
```diff
@@ -503,4 +505,21 @@ public void onSessionCreateFailure(Throwable t, int createFailureForSessionCount
...
+  public void testSessionNamePatterns() {
+    // Valid pattern for host session name
+    String host =
+        "projects/spanner-project/instances/spanner-instance/databases/test-db/sessions/abcd1234";
+    // Valid pattern for external host session name
+    String externalHost = "instances/default/databases/test-db/sessions/abcd1234";
+    try {
+      SessionId.of(host);
+      SessionId.of(externalHost);
+      // If no exceptions are thrown, the test will pass
```
Comment (krishnamoorthy-r) [inline google-cloud-spanner/src/test/java/com/google/cloud/spanner/SessionClientTests.java:520]: Is it possible to add assertEquas here to validate all the attributes instanceid, projectid and databaseid as set in both the scenarios?

**Review by krishnamoorthy-r — CHANGES_REQUESTED** [review]

**Review by krishnamoorthy-r — APPROVED** [review]

File: google-cloud-spanner/src/test/java/com/google/cloud/spanner/SessionClientTests.java (hunk 4 above)
Comment (sagnghos) [reply google-cloud-spanner/src/test/java/com/google/cloud/spanner/SessionClientTests.java:522]: added assert for session name.

**Review by rahul2393 — APPROVED** [review]

File: google-cloud-spanner/src/test/java/com/google/cloud/spanner/SessionClientTests.java [hunk 5]
```diff
@@ -503,4 +505,23 @@ public void onSessionCreateFailure(Throwable t, int createFailureForSessionCount
...
+        "projects/spanner-project/instances/spanner-instance/databases/test-db/sessions/abcd1234";
+    // Valid pattern for external host session name
+    String externalHost = "instances/default/databases/test-db/sessions/abcd1234";
+    try {
+      SessionId hostSession = SessionId.of(host);
+      assertEquals("abcd1234", hostSession.getName());
+      SessionId externalHostSession = SessionId.of(externalHost);
+      assertEquals("abcd1234", externalHostSession.getName());
+      // If no exceptions are thrown, the test will pass
+    } catch (IllegalArgumentException e) {
```
Comment (olavloite) [inline google-cloud-spanner/src/test/java/com/google/cloud/spanner/SessionClientTests.java:523]: nit: you can remove the entire try-catch block. If a (runtime) exception occurs during a test, then this test will be marked as failed with the error message. I.e. you will get the exact same behavior as you have now with the try-catch block, but with a bit less code and indentation.

**Review by olavloite — APPROVED** [review]
LGTM, with a couple of tiny nits.

**Pull request closed without merging**